### DISCUSSION TITLE
Is step-by-step playback capability available?

### DISCUSSION THREAD

**Discussion by antlaw** [excalidraw/excalidraw#8919] [General]
Hi there, 
I am new to excalidraw. Does excalidraw support a function that can step-by-step playback the whole drawing? If not, I can contribute to implement that. But can anyone give me a hint on how to start that?

Thanks very much creating excalidraw! It is cool!

Anthony

**Comment by Mrazator** [comment]
Hi, not at the moment, but we might have something similar in Excalidraw+ once we tackle versioning (it's on the roadmap, though super complex).

**Reply by antlaw** [reply]
HI Mrazator, 

Thanks very much for your prompt reply. 
I examined the api and structure of the project and come up with an initial idea of implementing the playback capability. Could you give me comment and let me know if there is something I have overlooked please? 
1. My idea is use onChange() + onPointerUpdate() to keep track of the screen and mouse pointer change. 
2. To playback the construction of the page, I will use the content captured to construct the page progressively. 

Anthony

**Reply by antlaw** [reply]
I utilize the onChange() function to track updates to elements, which allows me to playback changes using excalidrawAPI.updateScene(). The algorithm is straightforward:

- onChange(): Maintains a history of elements whenever changes occur.
- playback(): Iterates through the history and invokes excalidrawAPI.updateScene().

However, as the scene grows larger, calling excalidrawAPI.updateScene() requires more resources, leading to inefficiencies in memory and computation. Would it be feasible to implement an API like excalidrawAPI.updateSceneDelta() that updates only the delta changes?
